internal/gengapic/client_init.go

```diff
@@ -139,193 +139,405 @@ func (g *generator) clientOptions(serv *descriptor.ServiceDescriptorProto, servN
+		p("// Users should not Close this client.")
+		p("LROClient *lroauto.OperationsClient")
 		p("")
+		g.imports[pbinfo.ImportSpec{Name: "lroauto", Path: "cloud.google.com/go/longrunning/autogen"}] = true
+	}
 
-		p("// The x-goog-* metadata to be sent with each request.")
-		p("xGoogMetadata metadata.MD")
-		p("}")
+	if g.hasLROMixin() {
+		p("operationsClient longrunningpb.OperationsClient")
 		p("")
 
-		g.imports[pbinfo.ImportSpec{Path: "google.golang.org/grpc"}] = true
-		g.imports[pbinfo.ImportSpec{Path: "google.golang.org/grpc/metadata"}] = true
+		g.imports[pbinfo.ImportSpec{Name: "longrunningpb", Path: "google.golang.org/genproto/googleapis/longrunning"}] = true
 	}
 
-	// Client constructor
-	{
-		clientName := camelToSnake(serv.GetName())
-		clientName = strings.Replace(clientName, "_", " ", -1)
-
-		p("// New%sClient creates a new %s client.", servName, clientName)
-		p("//")
-		g.comment(g.comments[serv])
-		p("func New%[1]sClient(ctx context.Context, opts ...option.ClientOption) (*%[1]sClient, error) {", servName)
-		p("  clientOpts := default%sClientOptions()", servName)
-		p("")
-		p("  if new%sClientHook != nil {", servName)
-		p("    hookOpts, err := new%sClientHook(ctx, clientHookParams{})", servName)
-		p("    if err != nil {")
-		p("      return nil, err")
-		p("    }")
-		p("    clientOpts = append(clientOpts, hookOpts...)")
-		p("  }")
+	if g.hasIAMPolicyMixin() {
+
+		p("iamPolicyClient iampb.IAMPolicyClient")
 		p("")
-		p("  disableDeadlines, err := checkDisableDeadlines()")
-		p("  if err != nil {")
-		p("    return nil, err")
-		p("  }")
+
+		g.imports[pbinfo.ImportSpec{Name: "iampb", Path: "google.golang.org/genproto/googleapis/iam/v1"}] = true
+	}
+
+	if g.hasLocationMixin() {
+
+		p("locationsClient locationpb.LocationsClient")
 		p("")
-		p("  connPool, err := gtransport.DialPool(ctx, append(clientOpts, opts...)...)")
+
+		g.imports[pbinfo.ImportSpec{Name: "locationpb", Path: "google.golang.org/genproto/googleapis/cloud/location"}] = true
+	}
+
+	p("// The x-goog-* metadata to be sent with each request.")
+	p("xGoogMetadata metadata.MD")
+
+	p("}")
+	p("")
+
+	g.imports[pbinfo.ImportSpec{Path: "google.golang.org/grpc"}] = true
+	g.imports[pbinfo.ImportSpec{Path: "google.golang.org/grpc/metadata"}] = true
+
+	g.grpcClientUtilities(serv, servName, imp, hasRPCForLRO)
+}
+
+func (g *generator) grpcClientUtilities(serv *descriptor.ServiceDescriptorProto, servName string, imp pbinfo.ImportSpec, hasRPCForLRO bool) {
+	p := g.printf
+
+	clientName := camelToSnake(serv.GetName())
+	clientName = strings.Replace(clientName, "_", " ", -1)
+	lowcaseServName := lowerFirst(servName)
+
+	// Factory function
+	p("// New%sClient creates a new %s client based on gRPC.", servName, clientName)
+	p("//")
+	g.comment(g.comments[serv])
+	p("func New%[1]sClient(ctx context.Context, opts ...option.ClientOption) (*%[1]sClient, error) {", servName)
+	p("  clientOpts := default%[1]sGrpcCallOptions()", servName)
```

Comment:
I think we are conflating two different sets of options. One is the set of options provided to the factory function for use during client creation (specifically https://pkg.go.dev/google.golang.org/api/option#ClientOption) and call options used at call invocation (https://pkg.go.dev/github.com/googleapis/gax-go/v2#CallOption).

I think this should be `defaultClientOptions` ([this method](https://github.com/googleapis/google-cloud-go/blob/master/asset/apiv1/asset_client.go#L58-L68) used [here](https://github.com/googleapis/google-cloud-go/blob/master/asset/apiv1/asset_client.go#L191-L206)).

Reply:
If we want to separate the REST & gRPC default client options, another method may be necessary for that i.e. `defaultRestClientOptions` and `defaultGrpcClientOptions`. Although this is getting pretty duplicative and I wonder since `ClientOption` and `CallOption` are both transport agnostic interfaces, maybe we should just stuff all default options in one set and use for both 🤷🏻 . I'm ok with duplicating everything though if its safer because its just generated code.

Reply:
Good catch, you're absolutely right. I think it's better to keep the the `ClientOption` structs separate, which means two methods.

Reply:
I'm leaning more in favor of using distinct structures/methods with distinct names to disambiguate. It's just easier to keep track of and say "I'm writing REST surface stuff now, I need to use names that contain the 'Rest' moniker".